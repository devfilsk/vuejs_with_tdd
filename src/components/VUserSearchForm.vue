
<script>
export default {
    name: "UserSearchForm"
}
</script>

<template>
  
</template>

<style>

</style>