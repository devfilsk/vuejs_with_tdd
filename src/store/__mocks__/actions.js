import userFixture from '../../../tests/unit/fixture/user';
export default {
    SEARCH_USER: jest.fn().mockResolvedValue(userFixture)
}