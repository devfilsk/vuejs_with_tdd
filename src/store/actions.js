import api from '@/services/api'
import login from '@/services/login'
import * as types from './mutation-types';

export const SEARCH_USER = ({ commit }, { username }) => {
  return new Promise(async (resolve, reject) => {
    try {
      const user = await api.searchUser(username)
      console.log(user)
      commit(types.SET_USER, user)
      resolve(user)
    } catch(error) {
      reject(error)
    }
  })
}

export const LOGAR = ({ commit }) => {
  
}


