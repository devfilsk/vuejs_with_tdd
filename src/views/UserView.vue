<script>
import VUserSearchForm from '@/components/VUserSearchForm';
import VUserProfile from '@/components/VUserProfile';
import { mapState } from 'vuex';
import CardToggle from '@/components/CardToggle';
export default {
    name: "UserView",
    components: {
      VUserSearchForm,
      VUserProfile,
      CardToggle,
    },
    data(){
      return {
        searched: []
      }
    },
    methods: {
      searchUser(username) {
        this.$store.dispatch('SEARCH_USER', { username })
      },
      show(){
        console.log(user)
      }
    },
    computed: {
      ...mapState({
        user: 'user',
      }),
    }
}
</script>

<template>
  <div class="">
      <VUserSearchForm 
        @submitted="searchUser" 
      />
      <!-- <VUserProfile :user="user"/> -->
      <CardToggle :user="item" v-for="item in user" :key="item"/>
  </div>
</template>

<style>

</style>