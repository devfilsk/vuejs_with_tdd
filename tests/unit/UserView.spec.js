jest.mock('@/store/actions');
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import UserView from '@/views/UserView';
import VUserSearchForm from '@/components/VUserSearchForm';
import VUserProfile from '@/components/VUserProfile';
import initialState from '@/store/state';
import actions from '@/store/actions';
import userFixture from './fixture/user';

// Cria instancia local do vue que pode ser manipulada sem interferir a instancia global
const localVue = createLocalVue();
localVue.use(Vuex);

describe('UserView', () => {
    let state;

    // Faz essa função para evitar repetir este código em todos so testes
    const build = () => {
        // arrange - shallowMount renderiza apenas o primeiro nível de dependencia do componente
        //Segundo parametro serve para verificar se uma props está sendo passada ao componente filho
        const wrapper = shallowMount(UserView, {
            localVue,
            store: new Vuex.Store({ 
                state,
                actions,
            })
        });

        return {
            wrapper,
            userSearchForm: () => wrapper.find(VUserSearchForm),
            userProfile: () => wrapper.find(VUserProfile)
        };
    }

    // Inicia o stado antes dos testes serem executados 
    beforeEach(() => {
        jest.resetAllMocks(),
        state = { ...initialState }
    })

    it('render the component', () => {
        
        // arrange - shallowMount renderiza apenas o primeiro nível de dependencia do componente
        const { wrapper } = build();

        expect(wrapper.html()).toMatchSnapshot();
    })

    it('render main child components', () => {
        // arrange 
        const { userSearchForm, userProfile } = build();

        // assert
        // Verifica se os components existem
        expect(userSearchForm().exists()).toBe(true);
        expect(userProfile().exists()).toBe(true);
    })

    it('passed a binded user prop to user profile component', () => {
        // arrange
        state.user = userFixture;
        const { userProfile } = build();

        // wrapper.setData({
        //     user: {
        //         name: "Filipe Maciel"
        //     }
        // });

        // assert
        expect(userProfile().vm.user).toBe(state.user);
    })

    it('searches for a user when recerved "submitted"', () => {
        // usuario a ser enviado
        const expectUser = 'devfilsk';
        const { userSearchForm } = build();

        // act 
        userSearchForm().vm.$emit('submitted', expectUser);

        expect(actions.SEARCH_USER).toHaveBeenCalled();
        expect(actions.SEARCH_USER.mock.calls[0][1]).toEqual({ username: expectUser });
    })
})