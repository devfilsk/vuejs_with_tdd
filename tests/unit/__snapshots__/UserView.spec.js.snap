// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserView render the component 1`] = `
<div>
  <vusersearchform-stub></vusersearchform-stub>
</div>
`;
